### PR TITLE
The Atlas GccCheckerPlugins package as an external.

### PR DESCRIPTION
**Pull request by gartung**
Although the PR is based on the gcc7 branch it can be back ported to any branch with a compiler that supports plugins and c++11.

### CODE REVIEW

**Comment by gartung** [comment]
@smuzaffar @Dr15Jones

**Comment by gartung** [comment]
@smuzaffar 

When I try to run the plugin as compiled is fails with

cc1plus: error: cannot load plugin /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms2/lib/libchecker_gccplugins.so
   /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms2/lib/libchecker_gccplugins.so: undefined symbol: _Z19get_inner_referenceP9tree_nodePlS1_PS0_P12machine_modePiS5_S5_

If I link the plugin against libcp1plugin.so, it fails with

cc1plus: error: cannot load plugin /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms/lib/libchecker_gccplugins.so
   /home/gartung/CMSSW_10_2_0/external/slc7_amd64_gcc630/lib/libcp1plugin.so.0: undefined symbol: cp_binding_oracle

If I link the plugin against libcc1plugin.so, it fails with

cc1plus: error: cannot load plugin /home/gartung/cms-build/slc7_amd64_gcc700/external/gcc-checker-plugin/1.0-cms4/lib/libchecker_gccplugins.so
   /home/gartung/CMSSW_10_2_0/external/slc7_amd64_gcc630/lib/libcc1plugin.so.0: undefined symbol: c_binding_oracle

**Comment by gartung** [comment]
Nevermind. I was able to determine that set up a project area for gcc630 and I compiled the plugin with gcc700.

**Comment by gartung** [comment]
Once the gcc-checker-plugin toolfile is configure in a project area the gcc checker can be run with 

g++ -c -o /dev/null --std=c++14 -fplugin=$GCC_CHECKER_PLUGIN -fplugin-arg-libchecker_gccplugins-checkers=all $file

**Comment by gartung** [comment]
@Dr15Jones I took the twiki text from your email and reformatted it.
https://twiki.cern.ch/twiki/bin/view/Sandbox/PatrickGartungSandbox

**Comment by gartung** [comment]
@davidlange6 Are you opposed to this one too?

**Comment by davidlange6** [comment]
not specifically - presumably if we want this for more than a quick test, we should get it put into a proper distribution (add bonus, with a makefile or similar) rather than downloading all the root/fastjet etc etc etc sources that atlas uses..

File: gcc-checker-plugin-toolfile.spec [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+EOF_TOOLFILE
+export GCC_CHECKER_PLUGIN_ROOT
+
+# General substitutions
+perl -p -i -e 's|\@([^@]*)\@|$ENV{$1}|g' %{i}/etc/scram.d/*.xml
+
+%post
+%{relocateConfig}etc/scram.d/*.xml
+echo "GCC_CHECKER_PLUGIN_TOOLFILE_ROOT='$CMS_INSTALL_PREFIX/%{pkgrel}'; export GCC_CHECKER_PLUGIN_TOOLFILE_ROOT" > $RPM_INSTALL_PREFIX/%{pkgrel}/etc/profile.d/init.sh
+echo "setenv GCC_CHECKER_PLUGIN_TOOLFILE_ROOT '$CMS_INSTALL_PREFIX/%{pkgrel}'" > $RPM_INSTALL_PREFIX/%{pkgrel}/etc/profile.d/init.csh
```
Comment (smuzaffar) [inline gcc-checker-plugin-toolfile.spec]: @gartung , I don't think you need all this logic in this spec file. Just copy the one e.g. curl-toolfile.spec and update variables.

File: gcc-checker-plugin.spec [hunk 2]
```diff
@@ -0,0 +1,23 @@
+### RPM external gcc-checker-plugin 1.0
```
Comment (smuzaffar) [inline gcc-checker-plugin.spec]: @gartung , use version 2.0.9 here and replace 2.0.9 with %{realversion} in rest of the file.

**Comment by davidlange6** [comment]
the trouble is that 2.0.9 has nothing to do with the version number of this specific package being built..

**Comment by smuzaffar** [comment]
but at least it points to a tag in gitlab ... right?

**Comment by davidlange6** [comment]
no. It seems to be a version of a tarball full of sources from many projects

cd atlasexternals-2.0.9-671e8f2b1413ff87772ea5dc024abeb5ef6ee441
ls
APE			CheckerGccPlugins	FastJetContrib		HEPUtils		MCUtils			SoQt
Acts			Coin3D			GPerfTools		HepMCAnalysis		MKL			XRootD
BAT			Davix			Gdb			KLFitter		PyAnalysis		dSFMT
Blas			Eigen			Geant4			Lapack			Python			dcap
Boost			FFTW			GoogleTest		Lhapdf			ROOT			lwtnn
CLHEP			FastJet			HDF5			LibXml2			Simage			yampl

**Comment by gartung** [comment]
I created this PR for review. It should not be merged yet. I plan to create a repo under cms-externals with just CheckerGccPlugin. I also want to change the Altas references to CMS. Is there a way to mark this PR WIP or For Review. I would mark the codecompass PR the same way.

File: gcc-checker-plugin-toolfile.spec (hunk 1 above)
Comment (gartung) [reply gcc-checker-plugin-toolfile.spec]: Eventually you will need the path to the plugin library.

**Comment by gartung** [comment]
I put the CheckerGccPlugins source in its own repository here
https://github.com/gartung/CheckerGccPlugins
I added a Makefile and changed all of the ATLAS references to CMS. This includes the checker_macros.h header that needs to be included to use the pragma macros needed to activate the thread checkers. The header could be placed in CMSSW if that is easier.

**Comment by gartung** [comment]
Mostly unmodified source now in cms-external repo
https://github.com/cms-externals/CheckerGccPlugins

**Comment by gartung** [comment]
Create gcc-analyzer-c(xx)compiler.xml so gcc checker can be run with
scram b COMPILER=gcc-analyzer

**Comment by smuzaffar** [comment]
please test

File: gcc-checker-plugin-toolfile.spec [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+mkdir -p %{i}/etc/scram.d
+
+cat << \EOF_TOOLFILE >%i/etc/scram.d/gcc-checker-plugin.xml
+  <tool name="gcc-checker-plugin" version="@TOOL_VERSION@">
+    <client>
+      <environment name="GCC_CHECKER_PLUGIN_ROOT" default="@TOOL_ROOT@"/>
+      <environment name="INCLUDE"   default="$GCC_CHECKER_PLUGIN_ROOT/include"/>
+    </client>
+  </tool>
+  <runtime name="GCC_CHECKER_PLUGIN" default="$GCC_CHECKER_PLUGIN_ROOT/lib/libchecker_gccplugins.so"/>
```
Comment (smuzaffar) [inline gcc-checker-plugin-toolfile.spec]: do we need GCC_CHECKER_PLUGIN env variable?

**Comment by smuzaffar** [comment]
please test

File: gcc-checker-plugin-toolfile.spec (hunk 3 above)
Comment (gartung) [reply gcc-checker-plugin-toolfile.spec]: It is no longer needed.

**Pull request merged**